app/services/places-import.js

```diff
@@ -0,0 +1,93 @@
+/*
+ Copyright 2016 Mozilla
+
+ Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ this file except in compliance with the License. You may obtain a copy of the
+ License at http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software distributed
+ under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ specific language governing permissions and limitations under the License.
+ */
+
+import {
+  datascript,
+  mori,
+  helpers,
+} from 'datascript-mori';
+import * as sqlite from './user-agent-service/sqlite';
+
+const d = datascript.core;
+
+const { vector, conj } = mori;
+const { DB_ADD } = helpers;
+
+export class PlacesImporter {
+  constructor(placesDB) {
+    this.db = placesDB;
+    this.places = new Map();
+  }
+
+  async buildPlaceMap(limit = undefined) {
+    const placeQuery = `SELECT p.id, p.url, p.title, p.visit_count, p.last_visit_date, p.guid,
+    hv.visit_date
+    FROM moz_places AS p LEFT JOIN moz_historyvisits AS hv
+    WHERE p.hidden = 0 AND p.id = hv.place_id
+    ORDER BY p.id, hv.visit_date
+    ${limit ? ' LIMIT ${limit}' : ''}
+    `;
+    await this.db.each(placeQuery, [], (row) => {
+      let place = this.places.get(row.id);
+      if (!place) {
+        place = {
+          id: row.id,
+          url: row.url,
+          title: row.title,
+          visit_count: row.visit_count,
+          last_visit_date: row.last_visit_date,
+          guid: row.guid,
+          visits: [],
+        };
+        this.places.set(row.id, place);
+      }
+
+      // For now, build entire places tree in memory.  TODO: stream to the DS store.
+      place.visits.push(row.visit_date);
+    });
+  }
+
+  writePlacesTo(conn) {
+    this.places.forEach((place) => {
+      let assertions = vector(
+        vector(DB_ADD, -1, 'page/url', place.url),
+        vector(DB_ADD, -1, 'page/guid', place.guid));
+
+      if (place.title) {
+        assertions = conj(assertions,
+                   vector(DB_ADD, -1, 'event/title', -2),
+                   vector(DB_ADD, -2, 'event/instant', place.last_visit_date),
```

Comment:
This is the idiomatic way to do this, at least for page titles as we experience them now: a page has one title at a time, and it might change over time.

If you retract the most recent title — as you might when clearing recent history — the older title should transparently take over.

Doing it this way allows you to simply query for the page title, rather than having to manually find the most recent one. That's much more efficient, and the underlying data is still there if you need it (modulo DataScript limitations) — you can find all historical titles by querying across transactions.

If we need to manage multiple titles at the same time, then either (a) they'll be a different kind of thing — e.g., a title associated with a downloaded page summary record, or (b) we can alter our schema; e.g., associating a title with every individual visit, so we can show the right title in a history timeline.

Does that make sense?